src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -172,6 +239,88 @@ public Entry(object @object)
                 Object = @object;
             }
         }
+
+        private struct ChecksumSynchronizer
+        {
+            private readonly AssetService _assetService;
+
+            public ChecksumSynchronizer(AssetService assetService)
+            {
+                _assetService = assetService;
+            }
+
+            public async Task SynchronizeAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+            {
+                // get children of solution checksum object at once
+                var solutionChecksums = new HashSet<Checksum>();
+                var solutionChecksumObject = await _assetService.GetAssetAsync<SolutionChecksumObject>(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Info);
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Projects);
+
+                await _assetService.SynchronizeAssetsAsync(solutionChecksums, cancellationToken).ConfigureAwait(false);
+
+                // get children of project checksum objects at once
+                var projectChecksums = new HashSet<Checksum>();
```

Comment:
why not have one collection, and add everything to that, and then sync that all at once.  Why do you need the separate sets?